### DISCUSSION TITLE
Can't add unicode domain

### DISCUSSION THREAD

**Discussion by bystrup** [modoboa/modoboa#1870] [Q&A] [answered]
# Impacted versions

* Modoboa: 1.14.0
* installer used: Yes
* Webserver: Apache

# Steps to reproduce
Add new domain with unicode character: havmændene.dk

# Current behavior
HTTP 500 error in apache log:
"POST /admin/domains/new/ HTTP/2.0" 500 23726 "https://modoboa.bystrup.net/admin/domains/" "Mozilla/5.0 (Linux; Android 6.0; Nexus 5 Build/MRA58N) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/80.0.3987.149 Mobile Safari/537.36"

UWSGI debug 
[modoboa-add-unicode-domain.txt](https://github.com/modoboa/modoboa/files/4432538/modoboa-add-unicode-domain.txt)

# Expected behavior
Should add domain

# Video/Screenshot link (optional)

**Comment by neonknight** [comment, marked as answer]
the domain must be added as punycode: xn--havmndene-j3a.dk

however, modoboa should convert automatically. python3 code for that is:
```
import codecs
punycodename = codecs.encode(domainname, 'idna').decode()
```

**Comment by tonioo** [comment]
This issue looks fixed now, I guess it was a python2 problem with encoding. Anyway, using UTF8 characters in usernames or domains is a pretty complex subject. There is an opened issue regarding usernames (https://github.com/modoboa/modoboa/issues/1419).
Modoboa does allow the creation of domain and username including utf8 characters, but as @neonknight says, we must convert the domain name to punycode because SMTPUTF8 support is not good enough (many email clients are not compatible). Regarding local part, it's another story!
And I guess we would have to use the punycode version to login to Modoboa or dovecot.

**Reply by thinktrans** [reply]
Just to comment upon the "Local Part" of the Email and the considerations about the same. There was a systematic study undertaken about it and the observations and recommendations have been documented by the Universal Acceptance Steering Group. Here is the [link ](https://uasg.tech/download/uasg-028-considerations-for-naming-internationalized-email-mailboxes-en/) for those  who are interested.
